source/MaterialXGenShader/TypeDesc.cpp

```diff
@@ -47,10 +47,11 @@ const TypeDesc* TypeDesc::registerType(const string& name, unsigned char basetyp
         throw ExceptionShaderGenError("A type with name '" + name + "' is already registered");
     }
 
-    TypeDescPtr ptr = TypeDescPtr(new TypeDesc(name, basetype, semantic, size, editable, channelMapping));
-    map[name] = ptr;
+    std::unique_ptr<TypeDesc> uniquePtr(new TypeDesc(name, basetype, semantic, size, editable, channelMapping));
+    TypeDesc* rawPtr = uniquePtr.get();
+    map[name] = std::move(uniquePtr);
```

Comment:
Thanks for fixing this! It clearly should have been a unique_ptr from the start.

For longer term I think we could improve on encapsulating the ownership of TypeDesc instances better. Letting TypeDesc be a lightweight carrier of the pointer without exposing it, similar to string interning. To avoid the risk of API users taking ownership of them. But still keeping type compares efficient.

Reply:
Looks like an improvement to me.

Reply:
Thanks to David for putting this changelist together!

Niklas — Yes, we could consider an object like observer_ptr (https://en.cppreference.com/w/cpp/experimental/observer_ptr) for the future, which would represent all non-owning, client-facing pointers.  For now, though, I think returning a raw pointer to a const object is a fine solution.